Data/ByteString/Internal/Type.hs

```diff
@@ -530,6 +532,99 @@ packUptoLenChars len cs0 =
           go !p (c:cs)          = pokeFp p (c2w c) >> go (p `plusForeignPtr` 1) cs
       in go p0 cs0
 
+data S2W = Octets {-# UNPACK #-} !Int [Word8]
+         | Hichar {-# UNPACK #-} !Int {-# UNPACK #-} !Word
+
+#if MIN_VERSION_template_haskell(2,17,0)
+liftTyped :: forall a m. (TH.Lift a, TH.Quote m) => a -> TH.Code m a
+liftTyped = TH.liftTyped
+
+liftCode :: forall a m. m (TH.TExp a) -> TH.Code m a
+liftCode = TH.liftCode
+
+literalFromChar8 :: (MonadFail m, TH.Quote m) => String -> TH.Code m ByteString
+#else
+liftTyped :: forall a. TH.Lift a => a -> TH.Q (TH.TExp a)
+liftTyped = TH.unsafeTExpCoerce . TH.lift
+
+liftCode :: forall a. TH.Q TH.Exp -> TH.Q (TH.TExp a)
+liftCode = TH.unsafeTExpCoerce
+
+literalFromChar8 :: String -> TH.Q (TH.TExp ByteString)
+#endif
+-- | Template Haskell splice to convert string constants to compile-time
+-- ByteString literals.  Unlike the 'IsString' instance, the input string
+-- is validated to ensure that each character is a valid /octet/, i.e. is
+-- at most @0xFF@ (255).
+--
+-- Example:
+--
+-- > :set -XTemplateHaskell
+-- > ehloCmd :: ByteString
+-- > ehloCmd = $$(literalFromChar8 "EHLO")
+--
+literalFromChar8 "" = [||empty||]
+literalFromChar8 s = case foldr' op (Octets 0 []) s of
+    Octets n ws -> liftTyped (unsafePackLenBytes n ws)
+    Hichar i w  -> liftCode $ fail $ "non-octet character '\\" ++
+        show w ++ "' at offset: " ++ show i
+  where
+    op (fromIntegral . fromEnum -> !w) acc
```

Comment:
Could you please add a type signature to `op`? I'm having hard times to guess what `fromIntegral . fromEnum` is doing.